### DISCUSSION TITLE
openAudioSession while on a call crashes the app

### DISCUSSION THREAD

**Discussion by Faaatman** [Canardoux/flutter_sound#758] [Q&A]
I tried opening a session while on a call and it crashed the app.
That's how I opened a session
```Dart
Future<void> _initAudioPlayer() async {
    try {
      _audioPlayer!
          .openAudioSession(focus: AudioFocus.requestFocusAndStopOthers)
          .then((value) {
        setState(() {
          _audioPlayerReady = true;
        });
      });
    } catch (e) {
      print(e);
    }
  }
```
<details>
<summary>And these are the logs. apparently, try and catch didn't catch that throw?</summary>
<p>

```
[VERBOSE-2:ui_dart_state.cc(199)] Unhandled Exception: 'package:flutter_sound_platform_interface/flutter_sound_player_platform_interface.dart': Failed assertion: line 104 pos 12: 'findSession(aSession) == -1': is not true.
#0      _AssertionError._doThrowNew (dart:core-patch/errors_patch.dart:46:39)
#1      _AssertionError._throwNew (dart:core-patch/errors_patch.dart:36:5)
#2      FlutterSoundPlayerPlatform.openSession (package:flutter_sound_platform_interface/flutter_sound_player_platform_interface.dart:104:12)
#3      FlutterSoundPlayer._openAudioSession (package:flutter_sound/public/flutter_sound_player.dart:690:41)
#4      FlutterSoundPlayer.openAudioSession.<anonymous closure> (package:flutter_sound/public/flutter_sound_player.dart:651:17)
#5      FlutterSoundPlayer.openAudioSession.<anonymous closure> (package:flutter_sound/public/flutter_sound_player.dart:650:30)
#6      BasicLock.synchronized (package:synchronized/src/basic_lock.dart:31:24)
#7      FlutterSoundPlayer.openAudioSession (package:flutter_sound/public/flutter_sound_player.dart:650:17)

```

</p>
</details>
edit: removed useless logs

**Comment by Faaatman** [comment]
I have changed how I open the session which cleared the log.
I removed the `.then` to `await`
```Dart
Future<void> _initAudioPlayer() async {
    try {
      await _audioPlayer!
          .openAudioSession(focus: AudioFocus.requestFocusAndStopOthers);
      setState(() {
        _audioPlayerReady = true;
      });
    } catch (e) {}
  }
```
and I have [read that assert](https://dart.dev/guides/language/language-tour#assert) doesn't work in production which might be the cause of the problem?
I will try to build the app in production and check again.

**Comment by Larpoux** [comment]
Android or iOS or Web ?

Also I need the logs before the crash.

**Reply by Faaatman** [reply]
iOS. Sadly the crashes only happened when I pushed the app to Testflight. they don't happen when I'm debugging it only throws an error.
I think I know where the error is coming from but I'm not sure.
in the file `flutter_sound_player_platform_interface.dart` you check for the available session via assert `assert(findSession(aSession) == -1);` and assert doesn't run in production as I stated in the above comment.
The following code block is what was throwing the error.
```Dart
  void openSession(FlutterSoundPlayerCallback aSession,)
  {
    assert(findSession(aSession) == -1);

    for (var i = 0; i < _slots.length; ++i)
    {
      if (_slots[i] == null)
      {
        _slots[i] = aSession;
        return;
      }
    }
    _slots.add(aSession);
  }
```
I think since assert doesn't work in production, it's not checking thus crashing in the for loop? correct me if I'm wrong :)

**Reply by Larpoux** [reply]
> Sadly the crashes only happened when I pushed the app to Testflight. they don't happen when I'm debugging it only throws an error.


Maybe you can look to [#710](https://github.com/Canardoux/tau/issues/710)
There is a problem with Flutter FFmpeg : 

> You must also check your build.gradle :
>
> Remove any reference to repository jcenter()
> Add a reference to mavenCentral() if not already done

**Reply by Faaatman** [reply]
ah perfect then I might switch to the lite version instead.
and changed the center to mavenCentral.
will test and get back to you 👍🏼

**Reply by Faaatman** [reply]
I tried using `flutter_sound_lite: ^8.3.9` still the same issue.
and I got this error through sentry package.

```
MissingPluginException: MissingPluginException(No implementation found for method listen on channel flutter_sound_ffmpeg_event)
  File "platform_channel.dart", line 156, in MethodChannel._invokeMethod
  File "<asynchronous suspension>"
  File "platform_channel.dart", line 545, in EventChannel.receiveBroadcastStream.<fn>
```

**Reply by Larpoux** [reply]
> I tried using `flutter_sound_lite: ^8.3.9` still the same issue.

No, this is not the same issue. Now you have a problem because you call FFmpeg (either directly or indirectly) and FFmpeg is not available in the LITE flavor.

Actually, I suggest that you debug your App with FULL mode. Then, if you think that you can reduce the App size, you will change the favor.

**Reply by Faaatman** [reply]
My app size is already big and I want to make it lighter. To be honest I don't know where I'm using FFmpeg. I will investigate it more and get back to you.
Thank you for your help man :)

**Comment by Larpoux** [comment]
Also you can try to run your app in release mode without having to upload to TestFlight : 
```sh
flutter run ios --release
```

**Reply by Faaatman** [reply]
Oh yeah I forgot about that. Thanks!

**Comment by Larpoux** [comment]
```sh
flutter run --release
```
